lib/base/log.c

```diff
@@ -838,26 +974,15 @@ heim_audit_trail(heim_svc_req_desc r, heim_error_code ret, const char *retname)
     if (r->e_text && r->kv)
 	heim_audit_addkv(r, HEIM_SVC_AUDIT_VIS, "e-text", "%s", r->e_text);
 
-    nelem = r->kv ? heim_array_get_length(r->kv) : 0;
-    for (i=0, j=0; i < nelem; i++) {
-	heim_string_t s;
-	const char *kvpair;
-
-        /* We know these are strings... */
-	s = heim_array_get_value(r->kv, i);
-	kvpair = heim_string_get_utf8(s);
-
-	if (j < sizeof(kvbuf) - 1)
-	    kvbuf[j++] = ' ';
-	for (; *kvpair && j < sizeof(kvbuf) - 1; j++)
-	    kvbuf[j] = *kvpair++;
-    }
-    kvbuf[j] = '\0';
+    kvb.pos = 0;
+    heim_dict_iterate_f(r->kv, &kvb, audit_trail_iterator);
+    kvb.buf[kvb.pos] = '\0';
 
+    /* FIXME: do we actually want to log the reason prefix? */
     heim_log(r->hcontext, r->logf, 3, "%s %s %s %s %s%s%s%s",
              r->reqtype, retval, r->from,
              r->cname ? r->cname : "<unknown>",
              r->sname ? r->sname : "<unknown>",
-             kvbuf, r->reason ? " " : "",
+             kvb.buf, r->reason ? " reason=" : "",
```

Comment:
as far as I could tell, the `reason` would have had a `reason=` prefix by virtue of `heim_audit_vaddreason()` calling `fmtkv()`. This doesn't really strike me as aesthetically desirable but I didn't want to change the log format.